bpftrace: fix regressions with 0.20.4

martinetd opened the pull request:
## Description of changes

bpftrace v0.20.4 stopped extracting kheaders to /tmp because of security issues (which was the right things to do), but some things broke a bit so fix and add a test.

This replaces #319033

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

martinetd committed:
bpftrace: fix tcp*.bt/runqlat.bt scripts

tcp*.bt scripts tried to include glibc headers, which didn't work on
nixos because bpftrace tried to call clang-major with some options to
get the paths, but we never told it where to find it.
Give bpftrace the path directly instead of giving it a path to clang.

runqlat.bt doesn't work with glibc headers and requires kernel headers
which are no longer extracted, just provide its value.

Note the default clang output also had some clang-internal headers, but
these seem to be included anyway through libclang API without having to
give them here, so we only pass the glibc ones.

While here:
- also backport another bpftrace patch that made the tcp*.bt tools to
actually not require these includes
- and add tests that we can include these.

Co-authored-by: Arian van Putten <arian.vanputten@gmail.com>

Lyndeno commented:
Result of `nixpkgs-review pr 319836` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bpftrace (linuxKernel.packages.linux_4_19.bpftrace ,linuxKernel.packages.linux_4_19_hardened.bpftrace ,linuxKernel.packages.linux_5_10.bpftrace ,linuxKernel.packages.linux_5_10_hardened.bpftrace ,linuxKernel.packages.linux_5_15.bpftrace ,linuxKernel.packages.linux_5_15_hardened.bpftrace ,linuxKernel.packages.linux_5_4.bpftrace ,linuxKernel.packages.linux_5_4_hardened.bpftrace ,linuxKernel.packages.linux_6_1.bpftrace ,linuxKernel.packages.linux_6_1_hardened.bpftrace ,linuxKernel.packages.linux_6_6.bpftrace ,linuxKernel.packages.linux_6_6_hardened.bpftrace ,linuxKernel.packages.linux_6_8.bpftrace ,linuxKernel.packages.linux_6_8_hardened.bpftrace ,linuxKernel.packages.linux_6_9.bpftrace ,linuxKernel.packages.linux_6_9_hardened.bpftrace ,linuxKernel.packages.linux_hardened.bpftrace ,linuxKernel.packages.linux_latest_libre.bpftrace ,linuxKernel.packages.linux_libre.bpftrace ,linuxKernel.packages.linux_lqx.bpftrace ,linuxKernel.packages.linux_xanmod.bpftrace ,linuxKernel.packages.linux_xanmod_latest.bpftrace ,linuxKernel.packages.linux_xanmod_stable.bpftrace ,linuxKernel.packages.linux_zen.bpftrace ,linuxPackages-libre.bpftrace ,linuxPackages.bpftrace ,linuxPackages_4_19_hardened.bpftrace ,linuxPackages_5_10_hardened.bpftrace ,linuxPackages_5_15_hardened.bpftrace ,linuxPackages_5_4_hardened.bpftrace ,linuxPackages_6_1_hardened.bpftrace ,linuxPackages_6_6_hardened.bpftrace ,linuxPackages_6_8_hardened.bpftrace ,linuxPackages_6_9_hardened.bpftrace ,linuxPackages_hardened.bpftrace ,linuxPackages_latest-libre.bpftrace ,linuxPackages_latest.bpftrace ,linuxPackages_lqx.bpftrace ,linuxPackages_xanmod.bpftrace ,linuxPackages_xanmod_latest.bpftrace ,linuxPackages_xanmod_stable.bpftrace ,linuxPackages_zen.bpftrace)</li>
    <li>bpftrace.man (linuxKernel.packages.linux_4_19.bpftrace.man ,linuxKernel.packages.linux_4_19_hardened.bpftrace.man ,linuxKernel.packages.linux_5_10.bpftrace.man ,linuxKernel.packages.linux_5_10_hardened.bpftrace.man ,linuxKernel.packages.linux_5_15.bpftrace.man ,linuxKernel.packages.linux_5_15_hardened.bpftrace.man ,linuxKernel.packages.linux_5_4.bpftrace.man ,linuxKernel.packages.linux_5_4_hardened.bpftrace.man ,linuxKernel.packages.linux_6_1.bpftrace.man ,linuxKernel.packages.linux_6_1_hardened.bpftrace.man ,linuxKernel.packages.linux_6_6.bpftrace.man ,linuxKernel.packages.linux_6_6_hardened.bpftrace.man ,linuxKernel.packages.linux_6_8.bpftrace.man ,linuxKernel.packages.linux_6_8_hardened.bpftrace.man ,linuxKernel.packages.linux_6_9.bpftrace.man ,linuxKernel.packages.linux_6_9_hardened.bpftrace.man ,linuxKernel.packages.linux_hardened.bpftrace.man ,linuxKernel.packages.linux_latest_libre.bpftrace.man ,linuxKernel.packages.linux_libre.bpftrace.man ,linuxKernel.packages.linux_lqx.bpftrace.man ,linuxKernel.packages.linux_xanmod.bpftrace.man ,linuxKernel.packages.linux_xanmod_latest.bpftrace.man ,linuxKernel.packages.linux_xanmod_stable.bpftrace.man ,linuxKernel.packages.linux_zen.bpftrace.man ,linuxPackages-libre.bpftrace.man ,linuxPackages.bpftrace.man ,linuxPackages_4_19_hardened.bpftrace.man ,linuxPackages_5_10_hardened.bpftrace.man ,linuxPackages_5_15_hardened.bpftrace.man ,linuxPackages_5_4_hardened.bpftrace.man ,linuxPackages_6_1_hardened.bpftrace.man ,linuxPackages_6_6_hardened.bpftrace.man ,linuxPackages_6_8_hardened.bpftrace.man ,linuxPackages_6_9_hardened.bpftrace.man ,linuxPackages_hardened.bpftrace.man ,linuxPackages_latest-libre.bpftrace.man ,linuxPackages_latest.bpftrace.man ,linuxPackages_lqx.bpftrace.man ,linuxPackages_xanmod.bpftrace.man ,linuxPackages_xanmod_latest.bpftrace.man ,linuxPackages_xanmod_stable.bpftrace.man ,linuxPackages_zen.bpftrace.man)</li>
  </ul>
</details>

rvl reviewed (approved):
I can confirm that this PR fixes the kernel headers issue. Maybe clean up the patch filenames a bit. Nice work.

martinetd commented:
> I can confirm that this PR fixes the kernel headers issue. Nice work.

Thanks for the review!

> Maybe clean up the patch filenames a bit.

Got something in mind? The filenames come straight out of "git format-patch" (with numbers reset to 1 as half of these are going away in the 0.21.0 upgrade follow-up) as it's the kind of names I'm used to, but if you have a preference I'm open to anything

rvl commented:
> > Maybe clean up the patch filenames a bit.
> 
> Got something in mind? The filenames come straight out of "git format-patch" (with numbers reset to 1 as half of these are going away in the 0.21.0 upgrade follow-up) as it's the kind of names I'm used to, but if you have a preference I'm open to anything

It's not too important, I was just thinking of keeping the git repo tidy. If the patches were going to hang around for a while, I would remove the numbers (`git format-patch -N`) and even further shorten the resulting filenames. No need to change anything. Someone should merge this.

martinetd committed:
bpftrace: only print kernel headers cannot be found warning on error

need more time to figure how to deal with kheaders cleanly

martinetd commented:
Since it hasn't been merged yet I've updated the last patch to now show the warning all the time (the quick&dirty show with -v has been replaced with a more proper "show on error" that I've sent upstream)

Also updated file names along the way so that's done, but the patch files will go away with the 0.21.1 update anyway... 3/5 are already merged there and the other two will be updated to fetchpatch.
The problem is that they ran clang format between 0.20 and 0.21 so none of the patches apply directly and I had to make it files specific to 0.20...

Anyway, if this doesn't get in I can't open the update PR, so will poke the discourse ready for merge thread tomorrow when I have a minute

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1793

SuperSandro2000 commented on pkgs/by-name/bp/bpftrace/package.nix:
```diff
@@ -41,9 +41,21 @@ stdenv.mkDerivation rec {
   ];
 
+  patches = [
+    # https://github.com/bpftrace/bpftrace/pull/3243 (merged)
+    ./override-system-headers.patch
+    # https://github.com/bpftrace/bpftrace/pull/3152 (merged)
+    ./tcp-bt-no-includes.patch
+    # https://github.com/bpftrace/bpftrace/pull/3262 (merged)
+    ./runqlat-bt-no-includes.patch
+    # https://github.com/bpftrace/bpftrace/pull/3242 (merged)
+    ./kheaders-not-found-message-fix.patch
+    # https://github.com/bpftrace/bpftrace/pull/3265
+    ./kheaders-not-found-message-only-on-error.patch
```
Why are we not using fetchpatch to obtain them?

martinetd replied:
>  The problem is that they ran clang format between 0.20 and 0.21 so none of the patches apply directly and I had to make it files specific to 0.20...

martinetd replied:
(I'm switching to fetchpatch for the couple of patches that stay in https://github.com/NixOS/nixpkgs/commit/875e88e8c28ff4cb4ab9065fe01291249ce35700 for the bpftrace 0.21.1 upgrade -- but thought it'd make sense to only fix things so we can backport that to stable branch and only update for master; if it's less work we can just update everything... But the current situation is broken so it'd be great to merge something, whichever it is)

SuperSandro2000 replied:
Good to know, then we can obviously not fetch them. Just re-request a review from me when you are done.

martinetd replied:
I consider this PR done -- the way I see it the last bpftrace 0.20.4 had regressions due to no longer extracting ikheaders, so:
- this PR fixes just the regressions (the last upstream PR has been merged since so it's all done)
- once that PR is in we can
  - backport exactly this PR to 24.05 so the regressions get fixed there
  - update nixpkgs-unstable to bpftrace 0.21.1, including using patches as per upstream

If we consider that keeping 24.05 to the older version isn't worth it, then we can just skip this PR and update straight to bpftrace 0.21.1 here then backport that update to 24.05; I don't use 24.05 so I don't have a preference either way (I think the regressions ought to be fixed, I don't care if we fix it by backport to the freezed version or if we just update it)

... Writing this I realized we can also just retarget this PR to 24.05 and open a new one that just updates master, there's no real benefit of 24.05 work being a backport is there.

Anyway, all the dev is done (including the update to bpftrace 0.21.1 in my tree), just tell me what you'd like to merge :)

Mic92 replied:
@martinetd updating to 0.21.1 on stable sounds good to me.

martinetd replied:
@Mic92 Thanks!

I've opened https://github.com/NixOS/nixpkgs/pull/327053 to update master to 0.21.1.

Okay for the update to stable, I don't have a preference either way; will backport that new PR to stable directly then
